File: src/test/java/com/churchapp/service/MemberServiceTest.java
package com.churchapp.service;

import com.churchapp.entity.Member;
import com.churchapp.repository.MemberRepository;
import arrow.core.Either;
import org.apache.camel.ProducerTemplate;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class MemberServiceTest {
    
    @Mock
    private MemberRepository memberRepository;
    
    @Mock
    private ProducerTemplate camelProducer;
    
    @InjectMocks
    private MemberService memberService;
    
    private Member testMember;
    
    @BeforeEach
    void setUp() {
        testMember = new Member(
            UUID.randomUUID(),
            "John",
            "Doe",
            "john.doe@example.com",
            "555-1234",
            LocalDate.of(1990, 1, 1),
            "123 Main St",
            LocalDate.now(),
            null, // baptismDate
            true, // isActive
            null, // family
            null, // user
            LocalDateTime.now(), // createdAt
            null  // updatedAt
        );
    }
    
    @Test
    void createMember_ShouldReturnSuccessfulTry_WhenValidMember() {
        // Given
        when(memberRepository.save(any(Member.class))).thenReturn(testMember);
        
        // When
        Try<Member> result = memberService.createMember(testMember);
        
        // Then
        assertTrue(result.isSuccess());
        assertThat(result.get()).isEqualTo(testMember);
        verify(memberRepository).save(testMember);
        verify(camelProducer).sendBody("direct:memberCreated", testMember);
    }
    
    @Test
    void createMember_ShouldReturnFailure_WhenRepositoryThrowsException() {
        // Given
        when(memberRepository.save(any(Member.class))).thenThrow(new RuntimeException("Database error"));
        
        // When
        Try<Member> result = memberService.createMember(testMember);
        
        // Then
        assertTrue(result.isFailure());
        assertTrue(result.getCause() instanceof RuntimeException);
        assertEquals("Database error", result.getCause().getMessage());
    }
    
    @Test
    void getMemberById_ShouldReturnSome_WhenMemberExists() {
        // Given
        when(memberRepository.findById(1)).thenReturn(Optional.of(testMember));
        
        // When
        var result = memberService.getMemberById(1);
        
        // Then
        assertTrue(result.isDefined());
        assertEquals(testMember, result.get());
    }
    
    @Test
    void getMemberById_ShouldReturnNone_WhenMemberDoesNotExist() {
        // Given
        when(memberRepository.findById(1)).thenReturn(Optional.empty());
        
        // When
        var result = memberService.getMemberById(1);
        
        // Then
        assertTrue(result.isEmpty());
    }
    
    @Test
    void getMemberByEmail_ShouldReturnSome_WhenMemberExists() {
        // Given
        when(memberRepository.findByEmail("john.doe@example.com")).thenReturn(Optional.of(testMember));
        
        // When
        var result = memberService.getMemberByEmail("john.doe@example.com");
        
        // Then
        assertTrue(result.isDefined());
        assertEquals(testMember, result.get());
    }
    
    @Test
    void deleteMember_ShouldReturnSuccessfulTry_WhenMemberExists() {
        // Given
        doNothing().when(memberRepository).deleteById(1);
        
        // When
        Try<Void> result = memberService.deleteMember(1);
        
        // Then
        assertTrue(result.isSuccess());
        verify(memberRepository).deleteById(1);
    }
    
    @Test
    void deleteMember_ShouldReturnFailure_WhenRepositoryThrowsException() {
        // Given
        doThrow(new RuntimeException("Delete failed")).when(memberRepository).deleteById(1);

        // When
        Try<Void> result = memberService.deleteMember(1);

        // Then
        assertTrue(result.isFailure());
        assertTrue(result.getCause() instanceof RuntimeException);
        assertEquals("Delete failed", result.getCause().getMessage());
    }

    @Test
    void updateMember_ShouldReturnSuccessfulTry_WhenValidMember() {
        // Given
        testMember.setFirstName("Jane");
        when(memberRepository.save(any(Member.class))).thenReturn(testMember);

        // When
        Try<Member> result = memberService.updateMember(testMember);

        // Then
        assertTrue(result.isSuccess());
        assertEquals("Jane", result.get().getFirstName());
        verify(memberRepository).save(testMember);
    }

    @Test
    void updateMember_ShouldReturnFailure_WhenRepositoryThrowsException() {
        // Given
        when(memberRepository.save(any(Member.class))).thenThrow(new RuntimeException("Update failed"));

        // When
        Try<Member> result = memberService.updateMember(testMember);

        // Then
        assertTrue(result.isFailure());
        assertTrue(result.getCause() instanceof RuntimeException);
    }

    @Test
    void getAllMembers_ShouldReturnListOfMembers() {
        // Given
        java.util.List<Member> members = java.util.Arrays.asList(testMember);
        when(memberRepository.findAll()).thenReturn(members);

        // When
        var result = memberService.getAllMembers();

        // Then
        assertEquals(1, result.size());
        assertEquals(testMember, result.head());
    }

    @Test
    void searchMembers_ShouldReturnMatchingMembers() {
        // Given
        String searchTerm = "John";
        java.util.List<Member> members = java.util.Arrays.asList(testMember);
        when(memberRepository.findByFirstNameContainingIgnoreCaseOrLastNameContainingIgnoreCase(searchTerm, searchTerm))
            .thenReturn(members);

        // When
        var result = memberService.searchMembers(searchTerm);

        // Then
        assertEquals(1, result.size());
        assertEquals(testMember, result.head());
    }
}

File: src/test/kotlin/com/churchapp/service/MemberServiceTest.kt
package com.churchapp.service

import com.churchapp.entity.Member
import com.churchapp.repository.MemberRepository
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.api.fail
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mock
import org.mockito.Mockito.`any`
import org.mockito.Mockito.`verify`
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension
import java.time.LocalDate
import java.util.Optional
import java.util.UUID

@ExtendWith(MockitoExtension::class)
@DisplayName("MemberService Tests")
class MemberServiceTest {
    @Mock
    private lateinit var memberRepository: MemberRepository

    private lateinit var memberService: MemberService
    private lateinit var member: Member

    @BeforeEach
    fun setup() {
        memberService = MemberService(memberRepository)

        member =
            Member.builder()
                .firstName("John")
                .lastName("Doe")
                .email("john.doe@example.com")
                .phoneNumber("123-456-7890")
                .address("123 Main St")
                .dateOfBirth(LocalDate.of(1990, 1, 1))
                .membershipDate(LocalDate.now())
                .build()
    }

    @Test
    fun `should create member successfully`() {
        `when`(memberRepository.save(any(Member::class.java))).thenReturn(member)
        `when`(memberRepository.findByEmail(anyString())).thenReturn(null)

        val result = memberService.createMember(member)

        assertTrue(result.isRight())
        result.fold(
            { fail("Expected success but got error") },
            { savedMember ->
                assertNotNull(savedMember)
                assertEquals("John", savedMember.firstName)
                verify(memberRepository).save(any(Member::class.java))
            },
        )
    }

    @Test
    fun `should fail to create member when email already exists`() {
        // Create a complete existing member with all required fields
        val existingMember =
            Member.builder()
                .id(UUID.randomUUID())
                .email(member.email)
                .firstName("Existing")
                .lastName("User")
                .phoneNumber("123-456-7890")
                .address("123 Main St")
                .dateOfBirth(LocalDate.of(1990, 1, 1))
                .membershipDate(LocalDate.now())
                .build()

        println("Test: Existing member email: ${existingMember.email}")
        println("Test: Member being created email: ${member.email}")

        `when`(memberRepository.findByEmail(member.email)).thenReturn(existingMember)

        val result = memberService.createMember(member)

        println("Test: Result is ${if (result.isLeft()) "Left" else "Right"}: $result")

        assertTrue(result.isLeft())
        result.fold(
            { error ->
                println("Test: Error received: $error")
                assertNotNull(error)
                assertTrue(error is MemberError.ValidationError)
                assertEquals("Validation error: Email already exists", (error as MemberError.ValidationError).message)
            },
            { success ->
                println("Test: Unexpected success: $success")
                fail("Expected error but got success")
            },
        )
    }

    @Test
    fun `should find member by id successfully`() {
        val id = UUID.randomUUID()
        member = Member.builderFrom(member).id(id).build()
        `when`(memberRepository.findById(id)).thenReturn(Optional.of(member))

        val result = memberService.findById(id)

        assertTrue(result.isRight())
        result.fold(
            { fail("Expected success but got error") },
            { foundMember ->
                assertNotNull(foundMember)
                assertEquals("John", foundMember.firstName)
            },
        )
    }

    @Test
    fun `should fail to find member by non-existent id`() {
        val id = UUID.randomUUID()
        `when`(memberRepository.findById(id)).thenReturn(Optional.empty())

        val result = memberService.findById(id)

        assertTrue(result.isLeft())
        result.fold(
            { error ->
                assertNotNull(error)
                assertTrue(error is MemberError.NotFound)
            },
            { fail("Expected error but got success") },
        )
    }

    @Test
    fun `should update member successfully`() {
        val id = UUID.randomUUID()
        val existingMember = Member.builderFrom(member).id(id).build()
        val updatedMember = Member.builderFrom(member).id(id).firstName("Jane").build()

        `when`(memberRepository.findById(id)).thenReturn(Optional.of(existingMember))
        `when`(memberRepository.findByEmail(anyString())).thenReturn(null)
        `when`(memberRepository.save(any(Member::class.java))).thenReturn(updatedMember)

        val result = memberService.updateMember(id, updatedMember)

        assertTrue(result.isRight())
        result.fold(
            { fail("Expected success but got error") },
            { savedMember ->
                assertNotNull(savedMember)
                assertEquals("Jane", savedMember.firstName)
                verify(memberRepository).save(any(Member::class.java))
            },
        )
    }

    @Test
    fun `should fail to update non-existent member`() {
        val id = UUID.randomUUID()
        `when`(memberRepository.findById(id)).thenReturn(Optional.empty())

        val result = memberService.updateMember(id, member)

        assertTrue(result.isLeft())
        result.fold(
            { error ->
                assertNotNull(error)
                assertTrue(error is MemberError.NotFound)
            },
            { fail("Expected error but got success") },
        )
    }

    @Test
    fun `should get active members successfully`() {
        val members = listOf(member)
        `when`(memberRepository.findByIsActive(true)).thenReturn(members)

        val result = memberService.getActiveMembers()

        assertTrue(result.isRight())
        result.fold(
            { fail("Expected success but got error") },
            { memberList ->
                assertNotNull(memberList)
                assertEquals(1, memberList.size)
                assertEquals("John", memberList.first().firstName)
            },
        )
    }

    @Test
    fun `should deactivate member successfully`() {
        val id = UUID.randomUUID()
        val activeMember = Member.builderFrom(member).id(id).isActive(true).build()
        val deactivatedMember = Member.builderFrom(member).id(id).isActive(false).build()

        `when`(memberRepository.findById(id)).thenReturn(Optional.of(activeMember))
        `when`(memberRepository.save(any(Member::class.java))).thenReturn(deactivatedMember)

        val result = memberService.deactivateMember(id)

        assertTrue(result.isRight())
        result.fold(
            { fail("Expected success but got error") },
            { savedMember ->
                assertNotNull(savedMember)
                assertFalse(savedMember.isActive)
                verify(memberRepository).save(any(Member::class.java))
            },
        )
    }
}
